magic_bridge/ic/src/dip20_proxy/src/api/withdraw.rs

```diff
@@ -1,33 +1,52 @@
+use ic_cdk::export::candid::Principal;
 use ic_kit::{
     candid::{candid_method, Nat},
     ic,
     macros::update,
-    Principal,
 };
 
 use crate::{
     common::{
+        cap::insert_claimable_asset,
         dip20::Dip20,
+        magic::Magic,
         tera::Tera,
-        types::{EthereumAddr, TokendId, TxError, TxFlag, TxReceipt},
+        types::{ClaimableMessage, EthereumAddr, TokendId, TxError, TxFlag, TxReceipt},
     },
-    proxy::{ToNat, ERC20_ADDRESS_ETH, STATE, TERA_ADDRESS},
+    proxy::{ToNat, ERC20_ADDRESS_ETH, MAGIC_ADDRESS_IC, STATE, TERA_ADDRESS},
 };
 
 /// withdraw left over balance if burn/mint fails
 /// this will attempt to bridge the leftover balance
 /// todo withdraw specific balance
 #[update(name = "withdraw")]
 #[candid_method(update, rename = "withdraw")]
-pub async fn withdraw(token_id: TokendId, eth_addr: EthereumAddr, _amount: Nat) -> TxReceipt {
+pub async fn withdraw(
+    eth_contract_as_principal: TokendId,
```

Comment:
should we expect users to pass in the eth contract instead of canister tokenId?

Reply:
yes, same as `Burn`